pkg/kubectl/bind-apiservice/plugin/servicebindings.go

```diff
@@ -139,13 +141,182 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
+		_, err = fmt.Fprintf(b, "Create and modify provider owned %s on your cluster.\n", groupResource)
+	case donate && selectorName != "":
+		_, err = fmt.Fprintf(b, "Create and modify the following user-owned %s on your cluster referenced with:\n	name: \"%s\"\n", groupResource, claim.Selector.Name)
+	case !donate && selectorName != "":
+		_, err = fmt.Fprintf(b, "Create and modify the following provider owned %s on your cluster referenced with:\n	name: \"%s\"\n", groupResource, claim.Selector.Name)
+	}
+
+	return err
+}
+
+func writeConsumerCase(b *bytes.Buffer, groupResource string, claim kubebindv1alpha1.PermissionClaim) error {
+	var err error
+
+	adopt := claim.Adopt
+	selectorName := claim.Selector.Name
+
+	switch {
+	case adopt && selectorName != "":
+		_, err = fmt.Fprintf(b, "The provider wants to become owner of the following user-created %s referenced with:\n	name: \"%s\"\nThe provider will be able to access, modify and delete said objects on your cluster.\n", groupResource, claim.Selector.Name)
+	case !adopt && selectorName != "":
+		_, err = fmt.Fprintf(b, "The provider wants read access to the following user-created %s referenced with:\n	name: \"%s\"\nThe provider will not be able to modify or delete said objects.\n", groupResource, claim.Selector.Name)
+	case adopt && selectorName == "":
+		_, err = fmt.Fprintf(b, "The provider wants to become owner of all user-created %s.\nThe provider will be able to access, modify and delete said objects on your cluster.\n", groupResource)
+	case !adopt && selectorName == "":
+		_, err = fmt.Fprintf(b, "The provider wants read access to all user-created %s.\nThe provider will not be able to modify or delete said objects.\n", groupResource)
+	}
+
+	return err
+}
+func writeOnConflict(b *bytes.Buffer, claim kubebindv1alpha1.PermissionClaim) error {
+	var err error
+
+	if claim.OnConflict != nil {
+		switch {
+		case claim.OnConflict.ProviderOverwrites && claim.OnConflict.RecreateWhenConsumerSideDeleted:
+			_, err = b.WriteString("Conflicting objects will be overwritten and created objects will be recreated upon deletion.\n")
+		case claim.OnConflict.ProviderOverwrites:
+			_, err = b.WriteString("Conflicting objects will be overwritten and created objects will not be recreated upon deletion.\n")
+		case claim.OnConflict.RecreateWhenConsumerSideDeleted:
+			_, err = b.WriteString("Conflicting objects will not be overwritten and created objects will be recreated upon deletion.\n")
+		default: //Do nothing
+		}
+	}
+	return err
+}
+
+func writeUpdateClause(b *bytes.Buffer, claim kubebindv1alpha1.PermissionClaim) error {
+	var err error
+
+	if claim.Update == nil {
+		return nil
+	}
+
+	donate := false
+	if claim.Create != nil {
+		donate = claim.Create.Donate
+	}
+
+	owner := claim.Selector.Owner
+	switch {
+	case (owner == kubebindv1alpha1.Provider && !donate && claim.Update.Fields != nil):
+		fallthrough
+	case (owner == kubebindv1alpha1.Provider && donate && claim.Update.Preserving != nil):
```

Comment:
These expressions have a lot of repetition (e.g. repeating owner checks). Have you tried simplifying them? It could make sense to introduce helper variables with semantic meanings.

Reply:
I reworked and simplified this in 926bb5993775516f6dbfe28787228a9b11d95f92.